sdk_container/src/third_party/coreos-overlay/profiles/coreos/targets/generic/dev/profile.bashrc

```diff
@@ -0,0 +1,29 @@
+# Adapted from cross-boss. Only needed for the dev profile because we don't
+# build the SDK with sysroot flags and we don't install build tools to the
+# production image.
+cros_pre_pkg_preinst_strip_sysroot() {
+	local FILES FILE
+
+	# Gather all the non-binary files with the --sysroot or --with-sysroot flag.
+	mapfile -t -d '' < <(find "${D}" -type f -exec grep -lZEIe "--sysroot=${ROOT}\b|--with-sysroot=${EROOT}\b" {} +) FILES
+
+	# Only continue if there are any.
+	[[ ${#FILES[@]} -eq 0 ]] && return
+
+	einfo "Stripping sysroot flags from:"
+	for FILE in "${FILES[@]}"; do
+		einfo "  - ${FILE#${D}}"
+	done
+
+	# Carefully strip the sysroot flags.
+	local sedargs=( -i -r )
+	local flag
+	for flag in --{,with-}sysroot="${EROOT}"; do
```

Comment:
Should this be the following instead? To match the flags you check with `grep`?

```suggestion
	for flag in "--sysroot=${ROOT}" "--with-sysroot=${EROOT}"; do
```

Reply:
I think `ROOT` and `EROOT` are basically the same thing in our case, because we don't use eprefix, but still.

Reply:
It's funny how Flatcar is forcing me to rethink things I've been doing fine for years. I'd forgotten that I didn't use `ESYSROOT` here because it's not defined in the pkg phases. I'm not sure whether the dev image is prepared in the same `ROOT` location as the binary packages it's built from, but I'd rather assume it's not. That means I need to change this hook from `pre_pkg_preinst` to `post_src_install`, and that means it needs to apply to the generic profile, not just the dev profile. While I'm happy to make this change for cross-boss, I'm now leaning towards just making a symlink for Flatcar. What do you think?

Reply:
When you say "dev image" do you mean dev container image? Both the production image and the developer container image are assembled from binpkgs in a temporary ROOT in BUILD_DIR.

A symlink leaking the build sysroot into production artifacts would be my least favourite option, I think we used to have a symlink in the build sysroot to itself (`/build/amd64-usr/build/amd64-usr -> ../`) but we cleaned that up.

Are we still sure the sysroot-wrappers approach wasn't cleaner than all of these workarounds that we're having to add?

Reply:
Yes, the dev container image. I know they are both built in a `ROOT`, but I'm not sure whether that location is exactly the same for both images. Probably not. If not, then `ESYSROOT` will differ from `EROOT`, and the stripping therefore won't work if only applied when installing the binary packages.

The symlink wouldn't be in the binary packages, so it wouldn't appear in the production image. It would just be added to the dev container image before finalising it.

I think setting these variables and adding a single symlink only to the dev container image is probably still cleaner on balance, but I admit that I have been wavering here.  I would like it to be cleaner, not just for Flatcar, but for cross-boss as well, so this has inspired me to look further into alternatives.

The specs-based solution isn't viable without changes to gcc, which might be hard sell to upstream, so I came up with a much simpler gcc patch (about 5 new lines in one place) that doesn't use specs at all. With that, things literally just work with no special flags or other adjustments needed. It's not like Gentoo doesn't patch gcc already. I'm running it past the Gentoo toolchain folks, and I'm currently testing out a full `@system` build with cross-boss.

Reply:
> Yes, the dev container image. I know they are both built in a ROOT, but I'm not sure whether that location is exactly the same for both images. Probably not.

It happens to be the same ROOT for both, and they are built sequentially but we shouldn't depend on that.

> If not, then ESYSROOT will differ from EROOT, and the stripping therefore won't work if only applied when installing the binary packages.

I don't follow this conclusion. At a given time ESYSROOT should be equal to EROOT, from `man emerge`:

```
SYSROOT = [path]
    (...) The value must either be / or equal to ROOT. When cross-compiling, only the latter is valid.
```
Unless you mean that the value of ROOT will be different between binpkg generation and installation.

> The symlink wouldn't be in the binary packages, so it wouldn't appear in the production image. It would just be added to the dev container image before finalising it.

Oh, sorry I didn't clarify. To me the dev container image is a production artifact that is shipped to users.

> The specs-based solution isn't viable without changes to gcc, which might be hard sell to upstream, so I came up with a much simpler gcc patch (about 5 new lines in one place) that doesn't use specs at all.

What does the gcc patch look like?

Reply:
> I don't follow this conclusion. At a given time ESYSROOT should be equal to EROOT, from man emerge:

I wrote that line. :grinning: It actually doesn't work for three reasons. As you say, `ROOT` can change between building and installation. Also, Portage doesn't set `ESYSROOT` during the pkg phases, and it doesn't preserve its value in binary packages.

Here's the gcc patch. I'll probably add something to the man page too.

```diff
diff -Naur a/gcc/gcc.cc b/gcc/gcc.cc
--- a/gcc/gcc.cc	2024-08-01 23:34:33.525082176 +0100
+++ b/gcc/gcc.cc	2024-08-01 23:43:31.557156041 +0100
@@ -5527,6 +5527,16 @@
 	      "BINUTILS", PREFIX_PRIORITY_LAST, 0, 1);
   free (tooldir_prefix);

+  if (*cross_compile == '1' && !target_system_root_changed)
+    {
+      const char *esysroot = env.get("ESYSROOT");
+      if (esysroot && esysroot[0] != '\0' && strcmp(esysroot, "/") != 0 && strcmp(esysroot, target_system_root) != 0)
+	{
+	  target_system_root = esysroot;
+	  target_system_root_changed = 1;
+	}
+    }
+
 #if defined(TARGET_SYSTEM_ROOT_RELOCATABLE) && !defined(VMS)
   /* If the normal TARGET_SYSTEM_ROOT is inside of $exec_prefix,
      then consider it to relocate with the rest of the GCC installation
```

My `@system` build successfully completed. There were no traces of the sysroot flag.

Reply:
It's got the initial thumbs up from Sam at Gentoo. I'm now doing a two-phase Jenkins build with it to really put it through its paces. If that works, I'll press ahead and open a new PR to supersede this one.